### PR TITLE
Add insurance buffer

### PR DESCRIPTION
**Pull request by kumar-ish**
## Motivation

As per the Tracer Perpetual Swaps whitepaper, the insurance pool for a market is composed of two accounts: the system-owned buffer insurance account, and the public insurance account. This pull request implements the system-owned buffer insurance account.

## Changes
* Add insurance buffer account

* Change respective functions to account for the insurance buffer

Note: Scope of PR changed so as to not include the insurance withdrawal fee, as aspects of it are still under research and development

### CODE REVIEW

File: contracts/Insurance.sol [hunk 1]
```diff
@@ -68,11 +69,11 @@ contract Insurance is IInsurance, Ownable, SafetyWithdraw {
         updatePoolAmount();
         InsurancePoolToken poolToken = InsurancePoolToken(token);
 
-        // tokens to mint = (pool token supply / collateral holdings) * collaterael amount to stake
+        // tokens to mint = (pool token supply / collateral holdings) * collateral amount to stake
```
Comment (greenbookwebb) [inline contracts/Insurance.sol:62]: How are we defining collateral holdings? Is collateral holdings just collateral amount (public), or is it collateral amount (public) + buffer? 

It should be the second one

File: contracts/Insurance.sol [hunk 2]
```diff
@@ -99,21 +100,26 @@ contract Insurance is IInsurance, Ownable, SafetyWithdraw {
         uint256 wadTokensToSend =
             LibInsurance.calcWithdrawAmount(
```
Comment (greenbookwebb) [inline contracts/Insurance.sol]: Just confiming that this function only allows you to withdraw from publc (collateralAmount), and not from bufferAmount.

File: contracts/Insurance.sol [hunk 3]
```diff
@@ -124,7 +130,12 @@ contract Insurance is IInsurance, Ownable, SafetyWithdraw {
         int256 quote = (tracer.getBalance(address(this))).position.quote;
         if (quote > 0) {
             tracer.withdraw(uint256(quote));
-            collateralAmount = collateralAmount + uint256(quote);
+            if (collateralAmount > 0) {
+                collateralAmount = collateralAmount + uint256(quote);
+            } else {
+                // Pay to buffer if nothing in public insurance
+                bufferAmount = bufferAmount + uint256(quote);
+            }
```
Comment (greenbookwebb) [inline contracts/Insurance.sol]: Some of this is true, some is incorrect. 

Insurance funding rate payments are paid proportionally to the two pools. e.g. If insurance funding payment = 25 Quote, Insurance buffer account has 10 units of quote already in the pool (10% of total), public account has 90 units of quote (90% of total), Insurance buffer account increases by 25*0.1, and public account increases by 25*0.9. 

BUT, if neither pool has any quote collateral in their account (like maybe at genesis), it will go 100% to insurance buffer account.

File: contracts/Insurance.sol [hunk 4]
```diff
@@ -199,7 +214,7 @@ contract Insurance is IInsurance, Ownable, SafetyWithdraw {
 
         uint256 ratio =
             PRBMathUD60x18.div(
-                getPoolTarget() - collateralAmount,
+                getPoolTarget() - collateralAmount - bufferAmount,
```
Comment (greenbookwebb) [inline contracts/Insurance.sol]: unsure what this does.

**Review by greenbookwebb — CHANGES_REQUESTED** [review]

File: contracts/Insurance.sol (hunk 1 above)
Comment (CalabashSquash) [reply contracts/Insurance.sol:62]: I think, at the moment, collateralAmount is just the public pool. However, it could probably be renamed to publicCollateralAmount.

File: contracts/Insurance.sol (hunk 2 above)
Comment (CalabashSquash) [reply contracts/Insurance.sol]: Yeah, I think `calcWithdrawAmount` should only take public collat amount, as opposed to public + buffer, right?

File: contracts/Insurance.sol (hunk 3 above)
Comment (kumar-ish) [reply contracts/Insurance.sol:138]: Yep, that makes sense; will fix this up

Comment (CalabashSquash) [reply contracts/Insurance.sol]: Why do you say 250.1 and 250.9? Where have these numbers come from? EDIT: this is a markdown formatting thing lol

Comment (kumar-ish) [reply contracts/Insurance.sol:138]: Yep, that makes sense; can fix

File: contracts/Insurance.sol (hunk 4 above)
Comment (kumar-ish) [reply contracts/Insurance.sol]: This ratio is `(insurance_fund_target - insurance_fund_holdings)/(leveraged_notional_value)` in the formula

Comment (kumar-ish) [reply contracts/Insurance.sol]: `collateralAmount` is the amount of quote in the public insurance pool and `bufferAmount` is the amount of quote in the buffer insurance pool

File: contracts/Insurance.sol (hunk 3 above)
Comment (CalabashSquash) [reply contracts/Insurance.sol]: According to @kumar-ish , the buffer account only gets paid when
a) The public account is empty
b) People pay withdrawal fees

But you're saying it's proportional to pool holdings.

If your statement is correct, should this split be done as soon as insurance funding rate payments are made, or is it OK to make this calculation when `updatePoolAmount(...)` is called?

File: contracts/Insurance.sol (hunk 4 above)
Comment (CalabashSquash) [reply contracts/Insurance.sol]: This makes the calculation for funding rate equal to

`36523 * (10^11) * [(target - public collat amount - buffer collat amount) / leveraged notional value]`

File: contracts/Insurance.sol (hunk 3 above)
Comment (kumar-ish) [reply contracts/Insurance.sol]: This should be fixed now; also, discussed in-person with Ben and it should be okay to make the calculation when `updatePoolAmount()` is called

File: contracts/Insurance.sol (hunk 2 above)
Comment (kumar-ish) [reply contracts/Insurance.sol]: Ooh, you're definitely right; fixed now

File: contracts/Insurance.sol (hunk 3 above)
Comment (CalabashSquash) [reply contracts/Insurance.sol]: Yep it is, because `updatePoolAmount()` is called on any change to insurance pool balance. I was worried the ratios could be "outdated" by the time it is called, but this is not the case.

File: contracts/Insurance.sol [hunk 5]
```diff
@@ -142,24 +163,28 @@ contract Insurance is IInsurance, Ownable, SafetyWithdraw {
...
-        // What the balance will be after
-        uint256 difference = collateralAmount - amount;
-        if (difference < 10**18) {
+        if (bufferAmount + collateralAmount < 10**18) {
             // Once we go below one token, social loss is required
             // This calculation caps draining so pool always has at least one token
-            amount = collateralAmount - (10**18);
-            // Use new amount to compute difference again.
-            difference = collateralAmount - amount;
+            bufferAmount = 10**18;
```
Comment (CalabashSquash) [inline contracts/Insurance.sol]: I think this is wrong. You can't just arbitrarily set bufferAmount to 1. Where is this token coming from?

**Review by CalabashSquash — CHANGES_REQUESTED** [review]

File: contracts/Insurance.sol (hunk 5 above)
Comment (kumar-ish) [reply contracts/Insurance.sol]: This is making the making sure there is at least one token in the insurance pool, but it should probably be a part of the public pool. I'm gonna speak to Webb about this and see how it should be specc'd

Comment (greenbookwebb) [reply contracts/Insurance.sol]: yep, as discussed, get rid - no need for this. Only for public account

File: contracts/Insurance.sol [hunk 6]
```diff
@@ -199,7 +224,7 @@ contract Insurance is IInsurance, Ownable, SafetyWithdraw {
 
         uint256 ratio =
             PRBMathUD60x18.div(
-                getPoolTarget() - collateralAmount,
+                getPoolTarget() - collateralAmount - bufferAmount,
                 levNotionalValue
```
Comment (greenbookwebb) [inline contracts/Insurance.sol]: If `levNotionalValue == 0`, this will throw. This can happen naturally, or happen when `maxlev == 1`

Comment (greenbookwebb) [reply contracts/Insurance.sol]: @BenjaminPatch @kumar-ish

Comment (kumar-ish) [reply contracts/Insurance.sol]: There's a check for `levNotionalValue` on line 224; following block of code:
```
    if (levNotionalValue <= 0) {
        return 0;
    }
```

File: contracts/Insurance.sol (hunk 4 above)
Comment (greenbookwebb) [reply contracts/Insurance.sol]: What's this for though? Insurance Funding Rate?

**Review by greenbookwebb — APPROVED** [review]

File: contracts/Insurance.sol (hunk 4 above)
Comment (kumar-ish) [reply contracts/Insurance.sol]: Yep, that's right

File: contracts/Insurance.sol (hunk 5 above)
Comment (kumar-ish) [reply contracts/Insurance.sol]: Fixed (I think)

File: contracts/Insurance.sol [hunk 7]
```diff
@@ -137,29 +161,55 @@ contract Insurance is IInsurance, Ownable, SafetyWithdraw {
...
+            } else {
+                //
+                amount = bufferCollateralAmount;
+            }
+
+            // Drain buffer
+            bufferCollateralAmount = 0;
+        } else if (amount > bufferCollateralAmount) {
+            uint256 leftAfterDrain =
+                amount - bufferCollateralAmount - publicCollateralAmount;
```
Comment (CalabashSquash) [inline contracts/Insurance.sol]: Consider the example where
`publicCollateralAmount = 10`
`bufferCollateralAmount = 20`
`amount = 25`

then `amount > bufferCollateralAmount == true`
so `leftAfterDrain = 25 - 20 - 10` so underflow would occur.

It should make `leftAfterDrain = amount - bufferCollateralAmount`, then check if `amount > public`, if so drain all but 1. Then `amount = amountTakenFromBuffer + amountTakenFromPublic`, and set bufferCollateralAmount to =, and amountTakenFromPublic to be whatever it should be (min 1 token)

**Review by CalabashSquash — CHANGES_REQUESTED** [review]

File: contracts/Insurance.sol (hunk 7 above)
Comment (kumar-ish) [reply contracts/Insurance.sol]: Yes, you're right -- I believe the logic should be all _patched_ up now. I did some [tests](https://gist.github.com/kumar-ish/23ff05e1cdf52be6826e0f7ee4f500c0) in Python too, but feel free to attack it with edge cases

File: contracts/Insurance.sol [hunk 8]
```diff
@@ -68,17 +69,17 @@ contract Insurance is IInsurance, Ownable, SafetyWithdraw {
...
                 poolToken.totalSupply(),
-                collateralAmount,
+                publicCollateralAmount,
                 wadAmount
             );
 
         // mint pool tokens, hold collateral tokens
         poolToken.mint(msg.sender, tokensToMint);
-        collateralAmount = collateralAmount + wadAmount;
+        publicCollateralAmount = publicCollateralAmount + wadAmount;
```
Comment (greenbookwebb) [inline contracts/Insurance.sol:67]: What is wadAmount? Is it the collateral that is being deposited into the pool when a user mints pool tokens?

Comment (kumar-ish) [reply contracts/Insurance.sol:67]: Yes, that's right

File: contracts/Insurance.sol [hunk 9]
```diff
@@ -121,10 +123,32 @@ contract Insurance is IInsurance, Ownable, SafetyWithdraw {
...
+            publicCollateralAmount = publicCollateralAmount + payToCollateral;
+
+            // Amount to pay to buffer is the remainder
+            bufferCollateralAmount =
+                bufferCollateralAmount +
+                quote -
+                payToCollateral;
+        } else {
+            // Pay to buffer if nothing in public insurance
+            bufferCollateralAmount = bufferCollateralAmount + quote;
```
Comment (greenbookwebb) [inline contracts/Insurance.sol]: This doesn't look right. At the time a user withdraws from the public fund, the withdrawal fee goes entirely to the insurance buffer account.

File: contracts/Insurance.sol [hunk 10]
```diff
@@ -137,29 +161,46 @@ contract Insurance is IInsurance, Ownable, SafetyWithdraw {
...
+                // from the public insurance such that 1 token is left in the public buffer
+                amount =
+                    bufferCollateralAmount +
+                    publicCollateralAmount - 
+                    10**18;
+                publicCollateralAmount = 10**18;
+            } else {
+                // Take out what you need from the public pool; there's enough for there to be >= 1 token left
+                publicCollateralAmount = poolHoldings - amount;
+            }
```
Comment (greenbookwebb) [inline contracts/Insurance.sol]: this looks correct, but there's a lot going on here so I won't confirm - @BenjaminPatch to confirm

**Review by greenbookwebb — CHANGES_REQUESTED** [review]
Most of this looks good. Just got to fix up that bit with the withdrawal fee going entirely to the insurance buffer account.

File: contracts/Insurance.sol [hunk 11]
```diff
@@ -121,10 +123,32 @@ contract Insurance is IInsurance, Ownable, SafetyWithdraw {
...
-        int256 quote = (tracer.getBalance(address(this))).position.quote;
-        if (quote > 0) {
-            tracer.withdraw(uint256(quote));
-            collateralAmount = collateralAmount + uint256(quote);
+        uint256 quote =
+            uint256((tracer.getBalance(address(this))).position.quote);
+
+        tracer.withdraw(quote);
+
+        if (publicCollateralAmount > 0) {
```
Comment (CalabashSquash) [inline contracts/Insurance.sol:113]: Just confirming this is correct, and shouldn't be `bufferCollateralAmount`.

Comment (kumar-ish) [reply contracts/Insurance.sol:113]: Yes; only when someone has deposited into the public insurance account do funds start accruing proportionally to the buffer and public pools; otherwise, they all go to the public pool

File: contracts/Insurance.sol (hunk 10 above)
Comment (CalabashSquash) [reply contracts/Insurance.sol]: There is a lot going on but there are a decent amount of edge cases that need checking. I'm confident that this is good.

File: contracts/Insurance.sol (hunk 9 above)
Comment (kumar-ish) [reply contracts/Insurance.sol]: This is for updating the amount of buffer and public collateral tokens in the insurance pool. In order to optimise for gas, all insurance funding payments are made in the `TracerPerpetualSwaps` contract; when you actually need to deposit or withdraw from the insurance pool, you need access to those collateral tokens, so `updatePoolAmount` gets them from the perp swaps contract (which is why `tracer.withdraw(quote);` is called at the beginning). 

```
            // Pay to buffer if nothing in public insurance
```
is saying that if no one has deposited collateral into the insurance account, then any amount that the Tracer has paid out to the insurance contract (from insurance funding rate payments) belongs to the buffer account. If someone has (i.e. the if-statement branch, where the collateral in the insurance contract is greater than 0), you pay proportionally to the buffer and public account.

The withdraw fee is still under discussion so I decided to leave it out of scope for this PR.

**Review by CalabashSquash — APPROVED** [review]

**Review by raymogg — APPROVED** [review]
